packages/indicators/src/__tests__/builders/buildArithmetic/buildArithmetic.test.ts

```diff
@@ -0,0 +1,149 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { ArithmeticConfig, buildArithmetic } from '../../../builders/buildArithmetic';
+import { AnalyticValue } from '../../../types';
+import { createAggregator } from '../stubs';
+import { ANALYTIC_RESPONSE_FIXTURES } from './buildArithmetic.fixtures';
+
+const aggregationsByCode = (codes: string[]) =>
+  Object.fromEntries(codes.map(code => [code, ['FINAL_EACH_YEAR']]));
+
+describe('buildArithmetic', () => {
+  const aggregator = createAggregator(ANALYTIC_RESPONSE_FIXTURES);
+
+  it('throws an error if a data element referenced in the formula has no defined aggregation', async () => {
+    const config = { formula: 'A + B', aggregation: { A: 'MOST_RECENT' } };
+    return expect(buildArithmetic({ aggregator, config, fetchOptions: {} })).toBeRejectedWith(
+      /B.* has no aggregation defined/,
+    );
+  });
+
+  it('calls `aggregator.fetchAnalytics` with `fetchOptions`', async () => {
+    const config = {
+      formula: 'A + B',
+      aggregation: { A: ['MOST_RECENT'], B: ['MOST_RECENT'] },
+    };
+    const fetchOptions = { organisationUnitCodes: ['TO'] };
+    await buildArithmetic({ aggregator, config, fetchOptions });
+
+    expect(aggregator.fetchAnalytics).toHaveBeenCalledOnceWith(
+      expect.anything(),
+      fetchOptions,
+      expect.anything(),
+    );
+  });
+
+  describe('calculations for a single orgUnit/period combo', () => {
+    const testData: [string, ArithmeticConfig, AnalyticValue[]][] = [
+      [
+        'simple calculation - integer',
+        {
+          formula: 'A_TO_2019 + B_TO_2019',
+          aggregation: aggregationsByCode(['A_TO_2019', 'B_TO_2019']),
+        },
+        [{ organisationUnit: 'TO', period: '2019', value: 3 }],
+      ],
+      [
+        'simple calculation - float',
+        {
+          formula: 'A_TO_2019 / B_TO_2019',
+          aggregation: aggregationsByCode(['A_TO_2019', 'B_TO_2019']),
+        },
+        [{ organisationUnit: 'TO', period: '2019', value: 0.5 }],
+      ],
+      [
+        'complex calculation',
+        {
+          formula: '((A_TO_2019 + B_TO_2019) * C_TO_2019) / (D_TO_2019 - E_TO_2019)',
+          aggregation: aggregationsByCode([
+            'A_TO_2019',
+            'B_TO_2019',
+            'C_TO_2019',
+            'D_TO_2019',
+            'E_TO_2019',
+          ]),
+        },
+        [{ organisationUnit: 'TO', period: '2019', value: -9 }],
+      ],
+      [
+        'division with zero',
+        {
+          formula: 'A_TO_2019 / (A_TO_2019 + B_TO_2019 - C_TO_2019)',
+          aggregation: aggregationsByCode(['A_TO_2019', 'B_TO_2019', 'C_TO_2019']),
+        },
+        [{ organisationUnit: 'TO', period: '2019', value: Infinity }],
```

Comment:
`Infinity` is most probably not what we want here. I suggest we simply exclude all `Infinity` values from the results (no matter how they came to be: division with 0 or other means). Is this a bit hacky?

The other way would be skipping analytics that include a division with `0`, but we'll need some calculation metadata to be provided by `@beyondessential/arithmetic.runArithmetic`

Reply:
How else can you produce `Infinity`?

Reply:
From a quick investigation:

* Calculations including `Infinity` (eg `Infinity + 1`)
* `Math.log(0)`
* A number exceeds the upper limit of the floating point numbers

Neither of which seem applicable here

Reply:
Yeah happy to filter them out then!